### PR TITLE
[refactor] mailIssueCommentToParticipants

### PR DESCRIPTION
**Pull request by 6543**
utilize map

### CODE REVIEW

**Comment by 6543** [comment]
blocked by #10475 (if this is merged first a backport would be harder ...)

**Comment by guillep2k** [comment]
I don't know. I carefully considered maps for this when I did it. Every time you access a map entry a hash function is executed and [possibly a bucket must be allocated](https://dave.cheney.net/2018/05/29/how-the-go-runtime-implements-maps-efficiently-without-generics). Maps are slower to populate than slices. IMHO the current implementation is faster because it checks the map only one time per item _after_ it is allocated with the required size. Especially in the 90% of the cases where the list is quite short. I've even did some benchmarks, I believe.

I could be looking at this all wrong, though. 😅

File: services/mailer/mail_issue.go [hunk 1]
```diff
@@ -42,83 +42,88 @@ func mailIssueCommentToParticipants(ctx *mailCommentContext, mentions []int64) e
...
-	visited := make(map[int64]bool, len(unfiltered)+len(mentions)+1)
+	for _, id := range ids {
+		toNotify[id] = struct{}{}
+	}
 
 	// Avoid mailing the doer
-	visited[ctx.Doer.ID] = true
-	// Avoid mailing explicit unwatched
+	delete(toNotify, ctx.Doer.ID)
+  // Avoid mailing explicit unwatched
```
Comment (6543) [inline services/mailer/mail_issue.go]:
```suggestion
	// Avoid mailing explicit unwatched
```

**Comment by 6543** [comment]
> I don't know. I carefully considered maps for this when I did it. Every time you access a map entry a hash function is executed and [possibly a bucket must be allocated](https://dave.cheney.net/2018/05/29/how-the-go-runtime-implements-maps-efficiently-without-generics). Maps are slower to populate than slices. IMHO the current implementation is faster because it checks the map only one time per item _after_ it is allocated with the required size. Especially in the 90% of the cases where the list is quite short. I've even did some benchmarks, I believe.
> 
> I could be looking at this all wrong, though.

we get rid of https://github.com/go-gitea/gitea/pull/10478/files#diff-959cf0e72f197b979a8047af5f475157L113-L119 witch is an impruvement in my opinion ... but how did you make the benchmarks?

**Comment by guillep2k** [comment]
> 
> 
> we get rid of https://github.com/go-gitea/gitea/pull/10478/files#diff-959cf0e72f197b979a8047af5f475157L113-L119 witch is an impruvement in my opinion ... but how did you make the benchmarks?

Those two loops run through the data exactly one time; you may think they go over the same items several times, but they don't. A `for ... range` on a map would be slower than that.

As for testing, I don't recall exactly, but there's only one way I'd do it, and it's programming loops with static data and timing them running in the 100,000 of times. The database must be excluded from the benchmarks because it's a common part of and the slowest by several orders of magnitude, of course.

**Comment by 6543** [comment]
so if map dont add speed to it, I'll close this ... ?

**Comment by guillep2k** [comment]
I don't know, others might have a different opinion than me. Or consider your code better for other reasons (e.g. clarity).

File: services/mailer/mail_issue.go [hunk 2]
```diff
@@ -42,83 +42,88 @@ func mailIssueCommentToParticipants(ctx *mailCommentContext, mentions []int64) e
...
+	}
 	return nil
 }
 
-func mailIssueCommentBatch(ctx *mailCommentContext, ids []int64, visited map[int64]bool, fromMention bool) error {
+func mailIssueCommentBatch(ctx *mailCommentContext, ids []int64, fromMention bool) error {
 	const batchSize = 100
 	for i := 0; i < len(ids); i += batchSize {
-		var last int
+		job := make([]int64, batchSize)
```
Comment (6543) [inline services/mailer/mail_issue.go]:
```suggestion
		var job []int64
```

**Comment by 6543** [comment]
this got it's change

**Pull request closed without merging**